### PR TITLE
[DEV-2904 & 2787] Auto-populate foreign country on city selection, handling city deselection as well

### PR DESCRIPTION
**Pull request by MaxwellKendall**
**High level description:**
2 things addressed in this PR: 

1. Auto-selecting a country after foreign city selection
2. Handling city de-selection on back space:
- If city-selection resulted in auto-population of country selection, then deselect country

**Technical details:**
### LocationPickerContainer
- parseCountries changes `this.state.state`to now has property `autoPopulated`
- `this.selectEntity()` auto-selects country based when country selection is "All Foreign countries"
- All the rest of the updates are prettier updates (my bad!!)

### LocationPicker
- lots of logic in componentDidUpdate 
- Lots of unit tests for logic

**JIRA Ticket:**
[DEV-2904](https://federal-spending-transparency.atlassian.net/browse/DEV-2904)

The following are ALL required for the PR to be merged:
- [x] Code review
- [x] local demo confirming UX w/ Melissa
- [x] ~Elastic search re-indexed (no pr open yet)~ will be re-indexed during deploy, these changes do not break anything.

### CODE REVIEW

File: src/js/components/search/filters/location/EntityDropdownAutocomplete.jsx [hunk 1]
```diff
@@ -12,6 +12,7 @@ const propTypes = {
     placeholder: PropTypes.string,
```
Comment (MaxwellKendall) [inline src/js/components/search/filters/location/EntityDropdownAutocomplete.jsx]: This is for 2787.

File: src/js/components/search/filters/location/EntityDropdown.jsx [hunk 2]
```diff
@@ -13,6 +13,7 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import EntityDropdownList from './EntityDropdownList';
```
Comment (MaxwellKendall) [inline src/js/components/search/filters/location/EntityDropdown.jsx]: This is for 2787.

File: src/js/components/search/filters/location/LocationPicker.jsx [hunk 3]
```diff
@@ -51,39 +53,59 @@ export default class LocationPicker extends React.Component {
...
+        }
+
+        if (cityDeselected && this.props.state.autoPopulated) {
+            // city was deselected which auto populated state selection, so clear the state selection
+            this.props.selectEntity('state', defaultLocationValues.state);
+        }
+        else if (cityDeselected && this.props.country.autoPopulated) {
+            this.props.selectEntity('country', { code: "FOREIGN", name: "ALL FOREIGN COUNTRIES", autoPopulated: true });
         }
     }
```
Comment (MaxwellKendall) [inline src/js/components/search/filters/location/LocationPicker.jsx]: This is for 2787 and 2904; handling deselect after auto-populate on city selection.

File: src/js/containers/search/filters/location/LocationPickerContainer.jsx [hunk 4]
```diff
@@ -3,14 +3,19 @@
  * Created by Kevin Li 10/30/17
```
Comment (MaxwellKendall) [inline src/js/containers/search/filters/location/LocationPickerContainer.jsx]: The only methods changed in this file is selectEntity, parseStates, and parseCountries. The rest are meaningless prettier changes (sorry).

**Comment by plooksm** [comment]
Looked at the demo and confirmed that for city+state, when you start deleting the city, the state field resets. "United States" remains in the dropdown and does not reset as designed.  When filtering by a foreign city, the foreign country field resets to show "All Foreign Countries" when you start deleting the city.  When switching between countries or states, the city field resets.

**Comment by MaxwellKendall** [comment]
@ebdabbs and @jonhill13 adding a DNM label until the back end is ready. Dont think they have a PR yet for it.

**Pull request merged**